text/0000-slatepack.md

```diff
@@ -0,0 +1,391 @@
+
+    - `epk` is ephemeral public key `encode(X25519(ephemeral_secret, basepoint))`
+      - `encode(data)` is canonical base64 from RFC 4648 without padding.
+      - `ephemeral_secret` is `random(32)` and MUST be new for every new `message key`
+
+    - `emk` is encrypted message key `encrypt[HKDF[salt, label](X25519(ephemeral secret, public key))](message key)`
+      - `message key` is 128-bit master message key generated by each recipient
+      - `encrypt[key](plaintext)` is ChaCha20-Poly1305 from RFC 7539 with a zero nonce.
+      - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256.
+      - `salt` is `X25519(ephemeral secret, basepoint) || public key`
+      - `label` is "slatepack_v0.1"
+
+- `mac`
+  - A Message Authentication Code is generated on the `header` data which is all of the header values concatenated
+    - `encode(HMAC[HKDF["", "header"](message key)](header))`
+      - `encode(data)` is canonical base64 from RFC 4648 without padding.
+      - `HMAC[key](message)` is HMAC from RFC 2104 with SHA-256.
+      - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256.
+
+#### Payload (age Enryption)
+
+A binary serialized slate is encrypted according to the age encryption specification. The steps taken here follow age as closely as possible to avoid losing any security properties. An fairly well-review [age library in rust](https://github.com/str4d/rage/tree/master/age) is available to use for implementation.
+
+Any deviations in SlatePack encryption from the exact cryptography steps and decisions made in age are unintentional and should be corrected unless they are explicitly stated as a deviation from the cryptography decisions made by age.
+
+The expected deviations from the age specification are in the formatting and structure of the contents of the file output data (as opposed to the encryption technique itself or the actual binary encrypted payload).
+
+- `payload`
+  - `nonce || STREAM[HKDF[nonce, "payload"](message key)](plaintext)`
+    - `nonce` is `random(16)`
+    - `message key` is 128-bit master message key for each recipient
+    - `STREAM` is `ChaCha20-Poly1305` in 64KiB chunks and a nonce structure of 11 bytes of big endian counter, and 1 byte of last block flag (0x00 / 0x01)
+    - `HKDF[salt, label](key)` is 32 bytes of HKDF from RFC 5869 with SHA-256
+
+### Armor
+
+The payload that will be armored is a binary serialized `SlatePack` JSON object. Armor is `Framing` wrapped around a `SimpleBase58Check` encoded `Payload`.
+
+#### Framing
+
+Armor uses specific `Headers`, `Footers` and `Periods` as `Framing` to contain its `Payload`.
+
+- `Header`
+  - Supported Headers:
+    - `BEGINSLATEPACK`
+    - `ENDSLATEPACK X/Y`
+      - `Y` is the total number of message parts for a given multipart message and `X` is the number of the current partial message
+      - Only used if there is more than one message part
+  - Regex: `^[>\n\r\t ]*BEGINSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$`
```

Comment:
Do we really want to support `[>\n\r\t ]*` after `BEGINSLATEPACK`? I would expect it to just be:
`^[>\n\r\t ]*BEGINSLATEPACK[> ]?([1-9]\/[1-9])?$`

Reply:
Please see other regex comment. We definitely need to refine this a bit before ready to accept :+1:

Reply:
> Do we really want to support [>\n\r\t ]* after BEGINSLATEPACK?

In [saltpack](https://saltpack.org/armoring), which this is derived from, the motivation behind this is in order to be able to gracefully pick up headers in situations where email clients or forums garble up the text in messy reply chains, like so:
```
> BEGINSLATEPACK
>			
>  	1/3. 
```

I can see edge cases where supporting `[>\n\r\t ]*` can be useful. Are there any problems that could arise from supporting it?